README.md
# DoNotSend - hacking the DNS protocol

*Works on Windows and Linux*

The DNS protocol is conventionally used to ask for the IP address of a given website. Here it's used to send messages and retrieve other messages, instead of asking for a website IP address and retrieving its IP address.

## Disclaimer

This tool can be used to send messages by exploiting a flaw in the DNS protocol, but it could be used (as someone pointed out in [the reddit thread](https://www.reddit.com/r/Python/comments/l0ubhs/hacking_the_dns_protocol_to_use_it_as_a_messaging/)) to exfiltrate data from a network. I'm **not responsible for any misuse of the project**. Also note that it's most likely that your ISP is logging your DNS queries, thus it's not a 100% anonymous.

## Setup

* Python >= 3.7
* Scapy >= 2.4
  * if it isn't installed alongside scapy:
    * libpcap
* venv
  * Sometimes the `wheel` module is needed as well

```shell
apt install python3-venv
python3 -m venv venv/
source venv/bin/activate

pip3 install scapy
# if it fails because it couldn't build the wheel:
pip3 uninstall scapy && pip3 install wheel && pip3 install scapy

cd src
# nota: run as admin because it's binding on + sniffing port 53:
python3 server.py "interface" "host name"
# or you can start it with a configuration file:
python3 server.py config.ini
# if it complains about libpcap not installed, then:
apt install libpcap0.8-dev
```

## client

We can include arbitrary data in the hostname which the server then can interpret and execute/relay.
Thus we put our data in the qname section of the query, encoded using base32, without the padding (we can easily recalculate it).

The queries sent are TXT DNS queries, otherwise (because we answer with TXT DNS replies) the replies will get lost/deleted when transmitted by peers (yes you read correctly, Google can ask the DNS if it knows `crafted-domain.my_dns.domain.example.com`).

```bash
python3 client.py [my_dns.domain.example.com] "message here"
```

If no message is given, `hello world` is sent.

You can also use the `client.sh` version, relying only on `dig`, `base32` and `base64`, instead of using `scapy`.

## server

It receives queries and read the wanted "fake" hostname, decode the data put in the hostname as base32.

Then it replies through a DNS TXT reply, where the data is encoded as base64 without padding.

### Running the server

```shell
cd src
# needs to run as root because it is binding port 53
python3 server.py [interface, for example eth0 on linux] [my_dns.domain.example.com]
```

## Having other big DNS relay your queries and answers

In a few steps I was able to configure my NS provider to set myself up as my own DNS, to get to reply to the weird domains I need to communicate.

For this examples, let's say my server is named `example.com`.

1. I added a `A` entry for `dns.example.com`, pointing to `my server ip here`
1. In the DNS servers configuration, I already had `ns1.provider.com`, I added myself as a DNS server: `dns.example.com`, pointing to `my server ip here`
1. Then, just wait a bit (can be as long as 48 hours) and you're good to go

Now I just have to tell my client scripts to use the domain `dns.example.com` to send messages to it and it works like a charm, even when asking Google about it!

## Documentation

* [DNS packet structure](doc/DNSPacketStructure.md)
* [Design choices](doc/design.md)


CHANGELOG.md
# Changelog

## v0.0.4

### Added
- different users with the same tag (when registering) are now different in the message log
- possibility to log only a specific kind of DNS requests (on the server, through the configuration file)

### Changed
- answers TTL changed from 1024 (17 minutes ~) to 60 (1 minute)

## v0.0.2
### Added
- converter.py, to encode/decode ascii in base 32 and base 64 flawlessly
- packet.py to encapsulate a lot of dull work
- threaded udp socket server binded on port 53 otherwise we have an ICMP type 3 error (port unreachable, because nothing is binded to it)
- simple chat server, anonymizing the ip addresses, can receive commands to get the messages (/consult), otherwise just add the message to the queue
- error catching on subdomains decoding errors

### Changed
- now using subdomains of a main domain instead of the qname field, in case it's filtered
- the replies come in DNS reply answer field, as TXT
- the answer now includes the original question
- the IP packet is constructed with a TOS of 0x28 (normal priority, high throughput)
- TTL was added to the answer
- transaction id added to DNS layer (random, generated by client)
- random source port for UDP packet
- the client can now get a message as command line parameter
- fixing a bug on the decoders: they were adding padding when it shouldn't (payload size modulo 4 = 0)

## v0.0.1
### Added
- basic client sending message in DNS question qname field
- basic server reading message from DNS question qname field, answer in DNS reply answer field

src/server.py
#!/usr/bin/env python3

import binascii
import os
import socket
import sys
import threading
from configparser import ConfigParser
from typing import List

from scapy.layers.dns import DNSRR, dnstypes
from scapy.layers.inet import IP
from scapy.sendrecv import send, sniff

from converter import Content, Domain
from packet import Packet
from utils import DNSAnswer, get_ip_from_hostname, init_logger


def socket_server(ip: str):
    # bind UDP socket to port 53
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.bind((ip, 53))
    # and read until the end of the world
    while True:
        s.recvfrom(1024)
    s.close()


class Server:
    @staticmethod
    def from_file(filename: str):
        if not os.path.exists(filename):
            raise FileNotFoundError(filename)

        config = ConfigParser()
        config.read(filename)

        return Server(
            config["server"]["interface"],
            config["server"]["domain"],
            config["server"]["host_ip"],
            config,
        )

    def __init__(self, iface: str, domain: str, ip: str, config: ConfigParser = None):
        self.interface = iface
        self.host_ip = ip
        self.domain = domain
        self.config = config

        # subdomain => function(msg, ip, domains)
        self.subservers = {}

        self.logger = init_logger()

    def register(self, **subservers):
        self.subservers.update(subservers)

    def on_query(self, message: str, src_ip: str, domains: List[str]) -> str:
        if domains and self.subservers.get(domains[0]):
            return self.subservers[domains[0]](message, src_ip, domains)
        return "test"

    def _make_message(self, qname: str, content: str) -> DNSRR:
        return DNSRR(
            rrname=qname,
            rdata=Content.encode(content),
            type=DNSAnswer.Type.Text,
            ttl=int(self.config["packets"]["ttl"]) if self.config else 60,
        )

    def _make_txt(self, packet: Packet) -> Packet:
        try:
            subdomain, *domains = packet.subdomain_from_qname.split(".")
            domains = domains[::-1]
            data = Domain.decode(subdomain)
        except binascii.Error:
            # couldn't decode, drop the packet and do nothing
            self.logger.debug("Couldn't decode subdomain in %s", packet.qname)
            return

        return Packet.build_reply(
            {
                "src": self.host_ip,
                "dst": packet.src,
                "dport": packet.sport,
                "dns": {
                    "id": packet.id,
                    "question": packet.question,
                    "messages": [
                        self._make_message(
                            packet.qname, self.on_query(data, packet.src, domains)
                        ),
                    ],
                },
            },
            self.domain,
        )

    def _make_a(self, packet: Packet) -> Packet:
        if self.config is None:
            return

        # if we receive a DNS A query for a subdomain, answer it with an ip from
        # the configuration file
        qname = packet.qname[:-1]  # remove final '.'
        if qname in self.config.sections():
            return Packet.build_reply(
                {
                    "src": self.host_ip,
                    "dst": packet.src,
                    "dport": packet.sport,
                    "dns": {
                        "id": packet.id,
                        "question": packet.question,
                        "messages": [
                            DNSRR(
                                rrname=packet.qname,
                                rdata=self.config[qname]["ip"],
                                type=DNSAnswer.Type.HostAddr,
                                ttl=int(self.config[qname]["ttl"]),
                            ),
                        ],
                    },
                },
                self.domain,
            )

    def _dns_responder(self, pkt: IP):
        packet = Packet(pkt, self.domain)
        answer = None

        request_name = dnstypes[packet.question.qtype]

        if self.config["server"]["log"] == "*" or \
                self.config["server"]["log"] == request_name:
            self.logger.info(
                "[DNS %s] Source %s:%i - on %s",
                request_name,
                packet.src,
                packet.sport,
                packet.qname,
            )

        # reject every packet which isn't a DNS A/TXT query
        if packet.is_valid_dnsquery(
            "A", self.config["server"]["root"] if self.config else ""
        ):
            answer = self._make_a(packet)
        elif packet.is_valid_dnsquery("TXT"):
            answer = self._make_txt(packet)

        if answer is not None:
            send(answer.packet, verbose=0, iface=self.interface)

    def run(self):
        # bind a UDP socket server on port 53, otherwise we'll have
        # ICMP type 3 error as a client, because the port will be seen
        # as unreachable (nothing being binded on it)
        t = threading.Thread(target=socket_server, args=(self.host_ip,))
        t.start()

        self.logger.info(f"DNS sniffer started on {self.host_ip}:53")
        sniff(
            filter=f"udp port 53 and ip dst {self.host_ip}",
            prn=self._dns_responder,
            iface=self.interface,
        )

        t.join()


def main(**subservers):
    if len(sys.argv) != 2 and len(sys.argv) != 3:
        print(sys.argv)
        print("Usage: %s interface hostname" % sys.argv[0])
        print("       %s config_file.ini" % sys.argv[0])
        sys.exit(-1)

    server = None

    if len(sys.argv) == 3:
        ip = get_ip_from_hostname(sys.argv[2])
        if ip is None:
            print("Couldn't resolve IP from hostname, consider using a config file")
            sys.exit(-1)

        server = Server(sys.argv[1], sys.argv[2], ip)
    elif len(sys.argv) == 2:
        server = Server.from_file(sys.argv[1])

    if server is not None:
        server.register(**subservers)
        server.run()


if __name__ == "__main__":
    main()
